Add `unknown` and `never` to typecheck.md (#682)
Roblox/luau: docs/_pages/typecheck.md

local b = nil
```

> ## `unknown` type

Reviewer: Can these be sub-headings? (three #s)

Author: Yeah.

Author: Done, and renamed `Primitive types` heading to `Builtin types`.

`unknown` is also said to be the _top_ type, that is it's a union of all types.

```lua
local a: unknown = "hello world!"
local b: unknown = 5
local c: unknown = function() return 5 end
```

> Unlike `any`, `unknown` will not allow itself to inhabit into a different type!

Reviewer: ```suggestion
Unlike `any`, `unknown` will not allow itself to be used as a different type!
```

```lua
local function unknown(): unknown
    return if math.random() > 0.5 then "hello world!" else 5
end

local a: string = unknown() -- not ok
local b: number = unknown() -- not ok
local c: string | number = unknown() -- not ok
```

In order to turn a variable of type `unknown` into a different type, you must apply [type refinements](#type-refinements) on that variable.

```lua
local x = unknown()
if typeof(x) == "number" then
    -- x : number
end
```

## `never` type

`never` is also said to be the _bottom_ type, meaning there doesn't exist a value that inhabits the type `never`. In fact, it is the _dual_ of `unknown`. `never` is useful in many scenarios, and one such use case is when type refinements proves it impossible:

```lua
local x = unknown()
if typeof(x) == "number" and typeof(x) == "string" then
    -- x : never
end
```

## `any` type

> `any` is just like `unknown`, except that it allows itself to inhabit into a different type. Essentially, it's an opt-out from the type system entirely.

Reviewer: ```suggestion
`any` is just like `unknown`, except that it allows itself to be used as any type without further checks or annotations. Essentially, it's an opt-out from the type system entirely.
```
